src/treesaslicer.jl

```diff
@@ -0,0 +1,183 @@
+- `ntrials`, `βs` and `niters` are annealing parameters, doing `ntrials` indepedent annealings, each has inverse tempteratures specified by `βs`, in each temperature, do `niters` updates of the tree.
+- `sc_weight::Float64`: The weight for space complexity in the optimization. Default is 1.0.
+- `rw_weight::Float64`: The weight for read-write complexity in the optimization. Default is 0.2.
+- `fixed_slices::Vector{LT}`: A vector of fixed slices that should not be altered. Default is an empty vector.
+- `optimization_ratio::Float64`: The ratio used for optimization. Default is 2.0.
+
+# References
+- [Recursive Multi-Tensor Contraction for XEB Verification of Quantum Circuits](https://arxiv.org/abs/2108.05665)
+"""
+Base.@kwdef struct TreeSASlicer{RT,IT,LT} <: CodeSlicer
+    sc_target::RT = 30
+    βs::IT = 14:0.05:15
+    ntrials::Int = 10
+    niters::Int = 10
+    sc_weight::Float64 = 1.0
+    rw_weight::Float64 = 0.2
+    fixed_slices::Vector{LT} = []
+    optimization_ratio::Float64 = 2.0
+end
+
+function slice_tree(code::NestedEinsum, size_dict::Dict{LT,Int}; sc_target=30, βs=14:0.05:15, ntrials=10, niters=10, sc_weight=1.0, rw_weight=0.2, fixed_slices=LT[], optimization_ratio=2.0) where LT
+    ixs, iy = getixsv(code), getiyv(code)
+    ninputs = length(ixs)
+    if ninputs <= 2
+        return SlicedEinsum(LT[], code)
+    end
+
+    ###### Stage 1: preprocessing ######
+    labels = _label_dict(ixs, iy)  # map labels to integers
+    inverse_map = Dict([v=>k for (k,v) in labels])  # the inverse transformation, map integers to labels
+    log2_sizes = [log2.(size_dict[inverse_map[i]]) for i=1:length(labels)]   # use `log2` sizes in computing time
+    labels = _label_dict(code)
+
+    if ntrials <= 0
+        return SlicedEinsum(fixed_slices, code)
+    end
+
+    ###### Stage 2: computing ######
+    trees, tcs, scs, rws, slicers = Vector{ExprTree}(undef, ntrials), zeros(ntrials), zeros(ntrials), zeros(ntrials), Vector{Slicer}(undef, ntrials)
+    @threads for t = 1:ntrials  # multi-threading on different trials, use `julia -t 5 xxx.jl` for setting number of threads.
+        tree = _exprtree(code, labels)
+        slicer = Slicer(log2_sizes, Int[labels[l] for l in fixed_slices])
+        
+        treesa_slice!(tree, log2_sizes, slicer; βs=βs, niters=niters, sc_target=sc_target, sc_weight=sc_weight, rw_weight=rw_weight, optimization_ratio=optimization_ratio)
+        tc, sc, rw = tree_timespace_complexity(tree, log2_sizes)
+        @debug "trial $t, time complexity = $tc, space complexity = $sc, read-write complexity = $rw."
+        trees[t], tcs[t], scs[t], rws[t], slicers[t] = tree, tc, sc, rw, slicer
+    end
+
+    ###### Stage 3: postprocessing ######
+    best_tree, best_tc, best_sc, best_rw, best_slicer = first(trees), first(tcs), first(scs), first(rws), first(slicers)
+    for i=2:ntrials
+        if exp2(tcs[i]) + rw_weight * exp2(rws[i]) < exp2(best_tc) + rw_weight * exp2(rws[i])
```

Comment:
Do not copy code. It propagates bugs and makes the code hard to maintain. For example here, the same bug appears again, although it is a lagacy one.